Add test suite for PORT_PHY_ATTR flex counter feature

dhanasekar-arista opened the pull request:
### Description of PR
Add comprehensive test suite for PORT_PHY_ATTR flex counter feature to validate port phy attribute monitoring (RX_SIGNAL_DETECT, FEC_ALIGNMENT_LOCK, RX_SNR).

Validates complete data flow: CLI → CONFIG_DB → FLEX_COUNTER_DB → syncd → COUNTERS_DB

Summary: Implements end-to-end testing for PORT_PHY_ATTR flex counter feature

Fixes # (N/A - new feature test suite)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] New Test case
    - [x] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [x] 202505
- [x] 202511

### Approach
#### What is the motivation for this PR?
PORT_PHY_ATTR is a new flex counter feature enabling per-lane SERDES diagnostics. This test suite ensures:

Correct integration across all SONiC components (CLI, CONFIG_DB, orchagent, syncd, COUNTERS_DB)
Proper lane count handling for different port speeds (1/2/4/8 lanes)
Persistence across config reload and reboot
Interval configuration propagates correctly

#### How did you do it?
Have added the following tests,

test_port_phy_attr_enable_and_validate
test_port_phy_attr_interval_change
test_port_phy_attr_config_reload_persistence
test_port_phy_attr_reboot_persistence

#### How did you verify/test it?
validated by running the test on a TH512 platform

platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_enable_and_validate[moby102] PASSED                                                                                                                                  [ 25%]
platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_interval_change[moby102] PASSED                                                                                                                                      [ 50%]
platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_config_reload_persistence[moby102] PASSED                                                                                                                            [ 75%]
platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_reboot_persistence[moby102] PASSED                                                                                                                                   [100%]

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

dhanasekar-arista committed:
Add test suite for PORT_PHY_ATTR flex counter feature

Add comprehensive tests for port-phy-attr flex counter feature:
  - CLI and DB validation (CONFIG_DB, FLEX_COUNTER_DB, COUNTERS_DB)
  - Lane count verification for RX_SNR, FEC_ALIGNMENT_LOCK, RX_SIGNAL_DETECT
  - Interval change, config reload, and reboot persistence tests

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista committed:
fix import paths

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista commented:
@prgeor should we restrict this test to run only on asic-gen 'th5' ([ref](https://github.com/sonic-net/sonic-mgmt/blob/master/ansible/group_vars/sonic/variables#L20)).

I don't see any such check or restriction in SAI code.

dhanasekar-arista committed:
Add a new test to validate latch status

Updated the existing tests to match the new COUNTERS_DB schema for latch
state update.
Added a new test to verify latch status transitions (T->T*, F->F*) on link state changes

Verified the tests:
platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_enable_and_validate[moby205] PASSED                                                                               [ 20%]
platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_interval_change[moby205] PASSED                                                                                   [ 40%]
platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_config_reload_persistence[moby205] PASSED                                                                         [ 60%]
platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_reboot_persistence[moby205] PASSED                                                                                [ 80%]
------------------------------------------------------------------------------------ live log teardown ------------------------------------------------------------------------------------
20/01/2026 14:46:06 memory_utilization.check_memory_threshol L0064 WARNING| Skipping memory check for monit-memory_usage due to zero value

platform_tests/counterpoll/test_counterpoll_phy_attr.py::test_phy_latch_status_transition[moby205] PASSED                                                                           [100%]
------------------------------------------------------------------------------------ live log teardown ------------------------------------------------------------------------------------

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista committed:
Skip the test for non TH5 platforms

Also added minor bug fixes.

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

dhanasekar-arista committed:
Fixed pre-commit checks

Signed-off-by: dhanasekar-arista <dhanasekar@arista.com>

yxieca reviewed (dismissed):
AI agent on behalf of Ying. Reviewed; no issues found.

prajjwal-arista committed:
Changed SNR validation from int to float

    Changed SNR Validation from int to float as per PR -
    https://github.com/sonic-net/sonic-sairedis/pull/1854

Signed-off-by: Prajjwal Singh <prajjwal@arista.com>

mssonicbld commented:
/azp run

yxieca reviewed (changes requested):
Thanks for the PORT_PHY_ATTR suite — the counter parsing and clear-on-read latch handling are well done. One blocking item, one question, and a few non-blocking nits.

### Blocking

- `test_phy_latch_status_transition` shuts the fanout port down (`fanout.shutdown(port_info['fanout_port'])`) and only restores it with `fanout.no_shutdown(...)` near the end, with no `try`/`finally` in between. If any of the intermediate `poll_for_latch_status` calls asserts or times out, the fanout port is left administratively down, which breaks that DUT link for every subsequent test in the session (and potentially other modules on the same testbed). Please wrap the shutdown → ... → no_shutdown sequence in `try`/`finally` (restore in `finally`), or use a fixture that restores link state on teardown.

### Question

- `_port_config_cache` / `_port_oid_map_cache` are module-global and populated once, then reused after the config reload (`test_phy_config_reload_persistence`) and cold reboot (`test_phy_reboot_persistence`). Are the SAI port OIDs in `COUNTERS_PORT_NAME_MAP` guaranteed to be stable across a cold reboot on th5? On config reload they typically are; for cold reboot it can be ASIC-dependent. If they aren't guaranteed stable, the caches should be invalidated after the reboot/reload so `verify_counters_db_data` doesn't look up a stale OID. Just want to confirm this is intentional.

### Non-blocking nits

- `get_sample_ports_with_lane_counts`: `port_oid = port_oid_map[intf_name]` is a direct dict index. A port present in the PORT table but absent from `COUNTERS_PORT_NAME_MAP` (inactive/unmapped) would raise `KeyError`. Consider `.get()` + skip/resample.
- `random.sample(list(port_configs.keys()), sample_size)` is unseeded and unguarded — it raises `ValueError` if a SKU has fewer ports than `sample_size`, and the non-deterministic port choice makes failures harder to triage. Consider guarding the size and/or logging the chosen ports.
- `test_phy_config_reload_persistence` runs `config save -y` but never restores the original startup config. Even though it persists the box-default enabled state, a backup/restore (or avoiding `config save`) would prevent leaking startup-config changes into the rest of the session.
- `test_phy_interval_change` sets the interval to 5000/10000ms and doesn't restore the original poll interval, which then leaks into later tests that assume default polling. Consider restoring it on teardown.

AI-assisted review, reviewed by Ying Xie (@yxieca).

mssonicbld commented:
/azp run

mssonicbld commented:
This PR has backport request label(s) for branch(es): 202505,202511, but is missing required test information. Please make sure you tick the tested branch(es) in the **Tested branch** section and provide test evidence (e.g., 202505: &lt;test result&gt;) in the **Test result** section as well in your PR description.

<div align="right"><sub>---Powered by SONiC BuildBot</sub></div>

mssonicbld commented:
Hi, there are workflow run(s) waiting for approval, you may be first-time contributor. I will notify maintainers to help approve once PR is approved. Thanks!

<div align="right"><sub>---Powered by SONiC BuildBot</sub></div>

mssonicbld commented:
/azp run

mrinal-arista commented:
@yxieca Thanks for the detailed review — addressed the items as follows:
  - Wrapped the fanout shutdown/link-flap flow in `try/finally`, so the port is restored even if polling or validation fails.
  - Invalidated the port/OID caches after cold reboot; config reload retains the existing port objects, so the cache is kept for that path.
  - Port selection: sample-port selection validates the runtime `COUNTERS_PORT_NAME_MAP` entry and required PHY counter data before selecting a port.
  - Sampling now selects only eligible ports, handles cases with fewer eligible ports than requested, and logs the selected ports.
  - Added module-level `config_db.json` backup/restore cleanup to restore persistent PHY-counterpoll configuration changed by the suite, including `config save` and poll-interval changes.

mrinal-arista commented:
/azp run

mrinal-arista commented:
Hi @lolyu @StormLiangMS, could you please help check why the PR checks have not been triggered after my latest commit? The Azure Pipeline /azp run was not automatically triggered, so the PR is currently blocked. Thank you.

dhanasekar-arista commented:
> @yxieca Thanks for the detailed review — addressed the items as follows:
> 
> * Wrapped the fanout shutdown/link-flap flow in `try/finally`, so the port is restored even if polling or validation fails.
> * Invalidated the port/OID caches after cold reboot; config reload retains the existing port objects, so the cache is kept for that path.
> * Port selection: sample-port selection validates the runtime `COUNTERS_PORT_NAME_MAP` entry and required PHY counter data before selecting a port.
> * Sampling now selects only eligible ports, handles cases with fewer eligible ports than requested, and logs the selected ports.
> * Added module-level `config_db.json` backup/restore cleanup to restore persistent PHY-counterpoll configuration changed by the suite, including `config save` and poll-interval changes.

@yxieca all your comments are addressed, kindly review this.

dhanasekar-arista commented:
/azpw retry

mssonicbld commented:
Retrying failed(or canceled) jobs...

mssonicbld commented:
No Azure DevOps builds found for sonic-net/sonic-mgmt#21601.

dhanasekar-arista commented:
/azpw rerun

mssonicbld commented:
/azp run
